### PR TITLE
Improved handling for S4 `standardGeneric` & `nonstandardGenericFunction`

### LINKED ISSUE #50: S4 generic tests not being counted
Discovered when testing `rtables`, two S4 generics [`cell_values`](https://github.com/Roche/rtables/blob/4abe26c9ff010c3be72b7ffa288cc74d7af55c19/R/tt_pos_and_access.R#L1066) and [`value_formats`](https://github.com/Roche/rtables/blob/4abe26c9ff010c3be72b7ffa288cc74d7af55c19/R/tree_accessors.R#L963) were not being traced properly.

We need to add tests for registering of S4 generics, and handle any cases that might arise when indexing the associated srcrefs with any documented objects.

### PR DESCRIPTION
**Pull request by dgkf**
Closes #50

I gotta admit that S4's internals are quite involved and this was solved mostly heuristically. 

* I added a test that tests the behavior for `nonstandardGenericFunction` S4 generics, while a similar design in `rtables` produces a `standardGeneric`. Why `methods` chooses one or the other is not clear to me. 
* When methods are defined for internal generics they seem to use `<name>` aliases, whereas when they're for generics defined outside the package they use `<name>,<signature>-method` aliases. Treating it this way seems to address the issue, but I haven't found any documentation that makes this behavior clear to put my mind at ease that this is the right way to do this.

Nevertheless, this covers the case of a few edge cases in the `rtables` package that were not being appropriately covered. 

To confirm:

```sh
git clone git@github.com:Roche/rtables.git
cd covtracer
```

```r
remotes::install_local("../rtables")

options(covr.record_tests = TRUE)
cov <- covr::package_coverage("../rtables")
ttdf <- covtracer::test_trace_df(cov)

alias_has_test <- function(ttdf, alias) {
  alias_df <- ttdf[ttdf$alias == alias, ]
  length(na.omit(alias_df[["test_name"]])) > 0
}

stopifnot(
  alias_has_test(ttdf, "cell_values"),
  alias_has_test(ttdf, "value_formats")
)
```

### CODE REVIEW

File: R/obj_namespace_name.R [hunk 1]
```diff
@@ -45,3 +45,11 @@ obj_namespace_name.character <- function(x, ns) {
...
   env_ns_name(x$parent_env)
 }
+
+#' @exportS3Method
+obj_namespace_name.standardGeneric <- function(x, ns) {
+  methods::packageSlot(x)
+}
+
+#' @exportS3Method
+obj_namespace_name.nonstandardGenericFunction <- obj_namespace_name.standardGeneric
```
Comment (dgkf) [inline R/obj_namespace_name.R:55]: Specialized S4 generic function's originating namespace discovery

File: R/srcrefs.R [hunk 2]
```diff
@@ -139,17 +139,47 @@ srcrefs.R6ClassGenerator <- function(x, ..., srcref_names = NULL,
...
+
+  objs <- list(sr)
+  names(objs) <- srcref_names
+  objs
+}
+
+#' @exportS3Method
+#' @importFrom utils getSrcref
+#' @rdname srcrefs
+srcrefs.nonstandardGenericFunction <- srcrefs.standardGeneric
```
Comment (dgkf) [inline R/srcrefs.R:158]: specialized srcref discovery for S4 generic functions and binding of the namespace as a srcref attribute

File: R/srcrefs.R [hunk 3]
```diff
@@ -139,17 +139,47 @@ srcrefs.R6ClassGenerator <- function(x, ..., srcref_names = NULL,
...
+  ns <- env_ns_name(environment(x@.Data))
+
+  # if method is defined in same package as generic, use generic name as
+  # signature alias, otherwise use methods_info-style method alias name
+  if (generic_origin_ns == ns) {
+    signatures <- x@generic
+  } else {
+    # as produced by `methods:::.methods_info`
+    signatures <- paste0(x@generic, ",", paste0(x@defined, collapse = ","), "-method")
+  }
```
Comment (dgkf) [inline R/srcrefs.R:178]: Discovered along the way, generics that live outside the target package follow a different alias naming convention than those from within.

Comment (maksymiuks) [reply R/srcrefs.R:178]: Wow, that would be a real nightmare to debug had it become an issue at one point! Great job

**Comment by maksymiuks** [comment]
That looks really good! Awesome work!

Should we use an opportunity and remove `covr` from remote sources https://github.com/Genentech/covtracer/blob/main/DESCRIPTION#L40? `covr` was recently pushed to CRAN

**Review by maksymiuks — APPROVED** [review]

**Pull request merged**